app/controllers/kudos_controller.rb

```diff
@@ -57,6 +57,26 @@ def create
         end
       end
     end
+  rescue ActiveRecord::RecordNotUnique
+    # Uniqueness checks at application level (Rails validations) are inherently
+    # prone to race conditions. If we pass Rails validations but get rejected
+    # by database unique indices, use the usual duplicate error message.
+    #
+    # https://api.rubyonrails.org/v5.1/classes/ActiveRecord/Validations/ClassMethods.html#method-i-validates_uniqueness_of-label-Concurrency+and+integrity
+    respond_to do |format|
+      format.html do
+        flash[:comment_error] = ts("You have already left kudos here. :)")
+        redirect_to request.referer
+      end
+
+      format.js do
+        # TODO: AO3-5635 Clean up kudos error handling in JS.
```

Comment:
It looks like [AO3-5635](https://otwarchive.atlassian.net/browse/AO3-5635) is specifically for the error handling when javascript is disabled, but this is the code for when javascript is enabled. Using the keys of the hash and hard-coding the messages in JS isn't very i18n-compatible, but other than that I think the error messages have the correct behavior. Should there be a new issue for kudos error handling when JS is enabled?

Reply:
If we fix 5635 using [Rails 6's `i18n_customize_full_message`](https://blog.bigbinary.com/2019/04/22/rails-6-allows-to-override-the-activemodel-errors-full_message-format-at-the-model-level-and-at-the-attribute-level.html) we can remove all the hard-coded error messages in this controller and in JS.

I'll make a new issue for the refactoring if someone fixes 5635 before Rails 6 though.

Reply:
Maybe the issue for 5635 should be rewritten to reflect the fact that it should also change the javascript code, as well? Or this comment should be removed?

Reply:
I can remove the 5635 comments.